pkg/spdk/disk/types.go

```diff
@@ -0,0 +1,139 @@
+package disk
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"slices"
+
+	"github.com/pkg/errors"
+
+	commonTypes "github.com/longhorn/go-common-libs/types"
+	spdksetup "github.com/longhorn/go-spdk-helper/pkg/spdk/setup"
+	helperutil "github.com/longhorn/go-spdk-helper/pkg/util"
+
+	"github.com/longhorn/longhorn-spdk-engine/pkg/util"
+)
+
+type BlockDiskDriver string
+
+const (
+	BlockDiskDriverAuto       = BlockDiskDriver("auto")
+	BlockDiskDriverAio        = BlockDiskDriver("aio")
+	BlockDiskDriverNvme       = BlockDiskDriver("nvme")
+	BlockDiskDriverVirtioPci  = BlockDiskDriver("virtio-pci")
+	BlockDiskDriverVirtioScsi = BlockDiskDriver("virtio-scsi")
+	BlockDiskDriverVirtioBlk  = BlockDiskDriver("virtio-blk")
+)
+
+type BlockDiskSubsystem string
+
+const (
+	BlockDiskSubsystemVirtio = BlockDiskSubsystem("virtio")
+	BlockDiskSubsystemPci    = BlockDiskSubsystem("pci")
+	BlockDiskSubsystemNvme   = BlockDiskSubsystem("nvme")
+	BlockDiskSubsystemScsi   = BlockDiskSubsystem("scsi")
+)
+
+const (
+	UioDriverUioPciGeneric = "uio_pci_generic"
+)
+
+type BlockDiskType string
+
+const (
+	BlockDiskTypeDisk = BlockDiskType("disk")
+	BlockDiskTypeLoop = BlockDiskType("loop")
+)
+
+type DiskStatus struct {
+	Bdf          string
+	Type         string
+	Driver       string
+	Vendor       string
+	Numa         string
+	Device       string
+	BlockDevices string
+}
+
+func GetDiskDriver(diskDriver, diskPathOrBdf string) (BlockDiskDriver, error) {
+	if isBDF(diskPathOrBdf) {
+		return getDiskDriverForBDF(diskDriver, diskPathOrBdf)
+	}
+
+	return getDiskDriverForPath(diskDriver, diskPathOrBdf)
+}
+
+func getDiskDriverForBDF(diskDriver, bdf string) (BlockDiskDriver, error) {
+	executor, err := helperutil.NewExecutor(commonTypes.ProcDirectory)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to get the executor for disk driver detection")
+	}
+
+	output, err := spdksetup.GetDiskStatus(bdf, executor)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to get disk status for BDF %s", bdf)
+	}
+
+	var diskStatus DiskStatus
+	err = json.Unmarshal([]byte(output), &diskStatus)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to unmarshal disk status for BDF %s", bdf)
+	}
+
+	switch BlockDiskDriver(diskDriver) {
+	case BlockDiskDriverAuto:
+		diskPath := ""
+		if BlockDiskDriver(diskDriver) == BlockDiskDriverAuto && BlockDiskDriver(diskStatus.Driver) != UioDriverUioPciGeneric {
```

Comment:
`BlockDiskDriver(diskDriver) == BlockDiskDriverAuto` is duplicated

Reply:
Fixed. Thank you.